File: LeetCode/剑指Offer32-II从上到下打印二叉树.py
# https://leetcode.cn/problems/cong-shang-dao-xia-da-yin-er-cha-shu-ii-lcof/description/?envType=study-plan&id=lcof&plan=lcof&plan_progress=flnj2s3
# 2023年03月25日 17:07:31

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        if not root: return []
        Layer=[root]
        nextLayer=[]
        ret=[]
        while Layer or nextLayer:
            layerRet=[]
            while Layer:
                curr=Layer.pop(0)
                layerRet.append(curr.val)
                if curr.left: nextLayer.append(curr.left)
                if curr.right: nextLayer.append(curr.right)
            ret.append(layerRet)
            Layer=nextLayer
            nextLayer=[]
        return ret
            

File: LeetCode/剑指Offer32-II从上到下打印二叉树.go
// https://leetcode.cn/problems/cong-shang-dao-xia-da-yin-er-cha-shu-ii-lcof/description/?envType=study-plan&id=lcof&plan=lcof&plan_progress=flnj2s3
// 2023年03月25日 17:18:56

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func levelOrder(root *TreeNode) [][]int {
	if root == nil {
		return [][]int{}
	}
	ret := make([][]int, 0)
	layer := []*TreeNode{}
	nextLayer := []*TreeNode{}
	layer = append(layer, root)
	for len(layer) != 0 || len(nextLayer) != 0 {
		layerRet := []int{}
		for len(layer) != 0 {
			curr := layer[0]
			layer = layer[1:]
			layerRet = append(layerRet, curr.Val)
			if curr.Left != nil {
				nextLayer = append(nextLayer, curr.Left)
			}
			if curr.Right != nil {
				nextLayer = append(nextLayer, curr.Right)
			}
		}
		// fmt.Println(layerRet)
		ret = append(ret, layerRet)
		layer = nextLayer
		nextLayer = []*TreeNode{}
	}
	return ret
}